### PR TITLE
toml config, take 2

### PR DESCRIPTION
**Pull request by karmacoma-eth**
## New design principles

- names, types and default values defined in a plain Config dataclass
    - easy to parse, pickle and pass around, with well defined property names
    - Config dataclasses are immutable
- config file loaded separately through basic toml, applies overrides to the default Config
- the arg parser can be generated from the Config dataclass and has *no defaults,* it applies overrides to the config object again
- the arg parser can still be reused for contract level and test level dev docs, and should require no copy because it has no defaults (so it only adds overrides for explicit parameters)

## Advantages:

- no external dependencies
- the Config dataclass is the source of truth for parameter names, values, types and defaults
    - easy to go to the definition of a setting
- the override order is very clear and explicit
- Config objects are aware of their parents, so when you request a value from a child, it checks first in itself and falls back to its parent
    - 👍 no copies
    - 👍 explainable (you can print a test config and see the source for each parameter — default, file, CLI, contract or test devdocs)
- should be easy to *generate* a sample config file with up to date defaults too

### CODE REVIEW

**Comment by karmacoma-eth** [comment]
different take on https://github.com/a16z/halmos/issues/125

File: src/halmos/config.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+        group=solver,
+    )
+
+    solver_parallel: bool = arg(
+        help="run assertion solvers in parallel", global_default=False, group=solver
+    )
+
+    solver_threads: int = arg(
+        help="set the number of threads for parallel solvers",
+        global_default=os.cpu_count() or 1,
```
Comment (daejunpark) [inline src/halmos/config.py:338]: i don't remember why i put the previous default value `min(32, os.cpu_count() + 4)`, but this new one makes sense.

File: src/halmos/sevm.py [hunk 2]
```diff
@@ -1228,17 +1228,22 @@ def __len__(self) -> int:
...
-    def __init__(self, options: Dict) -> None:
+    def __init__(self, options: HalmosConfig) -> None:
         self.options = options
         self.logs = HalmosLogs()
         self.steps: Steps = {}
 
-        is_generic = self.options["storage_layout"] == "generic"
+        # init unknown calls
+        hex_string = options.uninterpreted_unknown_calls.strip()
+        self.unknown_calls: List[int] = [int(x, 16) for x in hex_string.split(",") if x]
```
Comment (daejunpark) [inline src/halmos/sevm.py:1243]: makes sense to force them be base 16

File: src/halmos/config.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+
+    loop: int = arg(
+        help="set loop unrolling bounds",
+        global_default=2,
+        metavar="MAX_BOUND",
+    )
+
+    width: int = arg(
+        help="set the max number of paths",
+        global_default=2**16,
```
[the remark this answers is not in the document]
Comment (daejunpark) [reply src/halmos/config.py:132]: we can also make width == 0 means infinite similar to depth.

**Review by daejunpark — APPROVED** [review]

File: src/halmos/config.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+        """
+
+        # look up value in current object
+        value = object.__getattribute__(self, name)
+        if value is not None:
+            return value
+
+        # look up value in parent object
+        parent = object.__getattribute__(self, "_parent")
+        if value is None and parent is not None:
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/config.py:386]: you're right!

File: src/halmos/sevm.py [hunk 5]
```diff
@@ -2098,12 +2097,11 @@ def finalize(ex: Exec):
                 insn = ex.current_instruction()
                 opcode = insn.opcode
 
-                max_depth = self.options.get("max_depth", 0)
-                if max_depth and step_id > max_depth:
+                if (max_depth := self.options.depth) and step_id > max_depth:
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py:2100]: that's right, hence the truthiness test first (alternatively we could give it a 0 default, may be cleaner)

File: src/halmos/config.py [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+        help="run tests in the given contract. Shortcut for `--match-contract '^{NAME}$'`.",
+        global_default=None,
+        metavar="CONTRACT_NAME",
+    )
+
+    match_contract: str = arg(
+        help="run tests in contracts matching the given regex. Ignored if the --contract name is given.",
+        global_default="",
+        metavar="CONTRACT_NAME_REGEX",
+    )
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/config.py:109]: good catch!

File: src/halmos/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/config.py:132]: capturing offline discussion: 2**64 felt like an oddly specific way to represent infinity. I thought that if we're going to have a limit, might as well make it a reasonable limit.

But you're right, let's make it 0 for infinite by default

File: src/halmos/config.py (hunk 1 above)
Comment (karmacoma-eth) [reply src/halmos/config.py:338]: yes I was always wondering why I was running with 24 threads when my machine supports 20 😅

File: src/halmos/sevm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daejunpark) [reply src/halmos/sevm.py:2100]: got it. agree with the default of 0.

File: src/halmos/config.py [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+        help="set loop unrolling bounds",
+        global_default=2,
+        metavar="MAX_BOUND",
+    )
+
+    width: int = arg(
+        help="set the max number of paths, 0 for unlimited",
+        global_default=0,
+        metavar="MAX_WIDTH",
+    )
```
Comment (daejunpark) [inline src/halmos/config.py:135]: nice! don't forget to update the width check in __main__.py

**Comment by karmacoma-eth** [comment]
I'll add a full halmos.toml example in a follow up

**Pull request merged**